### PR TITLE
Easier Postmessage Origin handling & Popup Build Script

### PR DESCRIPTION
**Pull request by dan-f**
This PR:

- Removes the need to statically template in the application origin
- Adds a distributed CLI script to more easily build the popup

### CODE REVIEW

**Comment by dan-f** [comment]
@RubenVerborgh this is ready for review!

File: README.md [hunk 1]
```diff
@@ -137,36 +137,22 @@ To use the popup login flow, you'll need a popup application running on a
 trusted domain which authenticates the user, handles redirects, and messages the
 authenticated session back to your application.
 
-Due to the possible redirects and the security model of `window.postMessage`,
-you'll need to build a static popup app bound to your application's origin.
+In order to tell the user they're logging into *your* app, you'll need to
+build a static popup bound to your application's name.
 
-Keeping this in mind, here's how to get things working.
+Keeping this in mind, it's pretty simple to build a popup for your app!
```
Comment (RubenVerborgh) [inline README.md]: I suppose they could even just edit the HTML at this point, it has become so easy :wink:

File: bin/solid-auth-client.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+const fs = require('fs')
+const path = require('path')
+
+const program = require('commander')
+
+const { version } = require('../package.json')
+
+// helpers
+
+function ns(fn) {
```
Comment (RubenVerborgh) [inline bin/solid-auth-client.js]: Maybe a comment about what this does?

File: bin/solid-auth-client.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  return function() {
+    fn.apply(
```
Comment (RubenVerborgh) [inline bin/solid-auth-client.js]: There's no need for `apply` anymore in ES6, just use the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator).

File: bin/solid-auth-client.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+function generatePopup(trustedName, filename = 'popup.html') {
+  log(`Generating "${filename}" with trusted app name "${trustedName}".`)
+
+  const templateFilename = path.resolve(
+    __dirname,
+    '..',
+    'dist-popup/popup.html'
+  )
+  if (!fs.existsSync(templateFilename)) {
+    const issueUrl = 'https://github.com/solid/solid-auth-client/issues'
```
Comment (RubenVerborgh) [inline bin/solid-auth-client.js]: maybe hoist all the way up the file?

File: bin/solid-auth-client.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+function ns(fn) {
+  return function() {
+    fn.apply(
+      null,
+      ['[solid-auth-client]'].concat(Array.prototype.slice.call(arguments))
+    )
+  }
+}
+
+const log = ns(console.log.bind(console))
```
Comment (RubenVerborgh) [inline bin/solid-auth-client.js]: No need to bind; `log` works independently (in Node, not the browser).

File: popup-app/index.js [hunk 6]
```diff
@@ -19,11 +23,45 @@ const idps = [
...
-    <IdpSelect idps={idps} />
-  ),
-  document.getElementById('app-container')
-)
+requestAppOrigin().then(appOrigin => {
+  ReactDOM.render(
+    window.location.hash ? (
+      <IdpCallback
+        appOrigin={appOrigin}
+        afterLoggedIn={() => setTimeout(window.close, 750)}
```
Comment (RubenVerborgh) [inline popup-app/index.js]: why the timeout?

File: popup-app/index.js [hunk 7]
```diff
@@ -19,11 +23,45 @@ const idps = [
...
+  storeAppOrigin(appOrigin)
+  return appOrigin
+}
+
+function getStoredAppOrigin() {
+  return getData(sessionStorage).then(data => data.appOrigin)
+}
+
+function storeAppOrigin(origin) {
+  return updateStorage(sessionStorage, data => ({
```
Comment (RubenVerborgh) [inline popup-app/index.js]: Is the return necessary? (see above)

File: popup-app/index.js [hunk 8]
```diff
@@ -19,11 +23,45 @@ const idps = [
...
+  if (appOrigin) {
+    return appOrigin
+  }
+  const request = client(window.opener, '*')
+  appOrigin = await request({
+    method: 'getAppOrigin',
+    args: []
+  })
+  storeAppOrigin(appOrigin)
+  return appOrigin
```
Comment (RubenVerborgh) [inline popup-app/index.js]: The `appOrigin` var is not necessary, given that `storeAppOrigin` already returns it.

File: popup-app/index.js [hunk 9]
```diff
@@ -19,11 +23,45 @@ const idps = [
...
+}
+
+function getStoredAppOrigin() {
+  return getData(sessionStorage).then(data => data.appOrigin)
```
Comment (RubenVerborgh) [inline popup-app/index.js]: I'd just use async/await consistently.

File: src/ipc.js [hunk 10]
```diff
@@ -123,29 +123,30 @@ export const server = (childWindow: window, childOrigin: string) => (
     if (!req) {
       return
     }
-    if (origin !== childOrigin) {
+    if (origin !== clientOrigin) {
       console.warn(
-        `SECURITY WARNING: solid-auth-client is listening for messages from ${childOrigin},` +
+        `SECURITY WARNING: solid-auth-client is listening for messages from ${clientOrigin},` +
```
Comment (RubenVerborgh) [inline src/ipc.js]: I'd explicitly say that the message has been rejected.

File: src/popup.js [hunk 11]
```diff
@@ -51,6 +55,16 @@ export const loginHandler = (
   }
 }
 
+export const appOriginHandler = (req: request): ?Promise<response> => {
+  const { id, method } = req
+  switch (method) {
```
Comment (RubenVerborgh) [inline src/popup.js]: Could be a simple `if` I suppose.

File: bin/solid-auth-client.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+const warn = ns(console.log.bind(console))
+
+// CLI
+
+program
+  .version(version)
+  .command('generate-popup <trusted-name> [filename]')
```
Comment (RubenVerborgh) [inline bin/solid-auth-client.js]: Why is <trusted-name> required? Also, I found the name confusing.

File: .env.demo.example [hunk 13]
```diff
@@ -1 +0,0 @@
-POPUP_URI=http://localhost:8081/popup.html
```
Comment (RubenVerborgh) [inline .env.demo.example]: Some mentions of the .env files are left in the README.

**Pull request merged**